src/test/net_tests.cpp

```diff
@@ -677,8 +682,80 @@ BOOST_AUTO_TEST_CASE(ipv4_peer_with_ipv6_addrMe_test)
     BOOST_CHECK(1);
+
+    // Cleanup, so that we don't confuse other tests.
+    {
+        LOCK(cs_mapLocalHost);
+        mapLocalHost.erase(addr_in_mapLocalHost);
+    }
 }
 
+BOOST_AUTO_TEST_CASE(advertise_local_port)
+{
+    // Test that AdvertiseLocal() properly self-advertises us:
+    //
+    // 1. AdvertiseLocal() calls GetLocalAddress() which returns an address that is not routable.
+    // 2. AdvertiseLocal() overrides the address with whatever the peer has told us he sees us as.
+    // 2.1. For inbound connections we must override both the address and the port.
+    // 2.2. For outbound connections we must override only the address.
+
+    // Pretend that we bound to this port.
+    const uint16_t bind_port = 20001;
+    gArgs.ForceSetArg("-bind", strprintf("3.4.5.6:%u", bind_port));
+
+    // Our address:port as seen from the peer.
+    in_addr peer_us_addr;
+    peer_us_addr.s_addr = htonl(0x02030405);
+    const CAddress peer_us{CService{peer_us_addr, 20002}, NODE_NETWORK};
+
+    // Create a peer with a routable IPv4 address (outbound).
+    in_addr peer_out_in_addr;
+    peer_out_in_addr.s_addr = htonl(0x01020304);
+    CNode peer_out{0,
+                   NODE_NETWORK,
+                   0,
+                   INVALID_SOCKET,
+                   CAddress{CService{peer_out_in_addr, 8333}, NODE_NETWORK},
+                   0,
+                   0,
+                   CAddress{},
+                   std::string{},
+                   ConnectionType::OUTBOUND_FULL_RELAY};
+    peer_out.fSuccessfullyConnected = true;
+    peer_out.SetAddrLocal(peer_us);
+
+    AdvertiseLocal(&peer_out);
+
+    // Without the fix peer_us:8333 ends up in vAddrToSend instead of the proper peer_us:bind_port.
+    const CAddress expected{CService{peer_us_addr, bind_port}, NODE_NETWORK};
+    BOOST_CHECK(std::any_of(peer_out.vAddrToSend.begin(), peer_out.vAddrToSend.end(),
+                            [&expected](const CAddress& cur) { return cur == expected; }));
+
+    // Create a peer with a routable IPv4 address (inbound).
+    in_addr peer_in_in_addr;
+    peer_in_in_addr.s_addr = htonl(0x05060708);
+    CNode peer_in{0,
+                  NODE_NETWORK,
+                  0,
+                  INVALID_SOCKET,
+                  CAddress{CService{peer_in_in_addr, 8333}, NODE_NETWORK},
+                  0,
+                  0,
+                  CAddress{},
+                  std::string{},
+                  ConnectionType::INBOUND};
+    peer_in.fSuccessfullyConnected = true;
+    peer_in.SetAddrLocal(peer_us);
+
+    AdvertiseLocal(&peer_in);
+
+    // Without the fix peer_us:8333 ends up in vAddrToSend instead of the proper
+    // peer_us:peer_us.GetPort().
+    BOOST_CHECK(std::any_of(peer_in.vAddrToSend.begin(), peer_in.vAddrToSend.end(),
+                            [&peer_us](const CAddress& cur) { return cur == peer_us; }));
+
+    gArgs.ForceSetArg("-bind", "");
```

Comment:
would be good to remove the global here, but that can be done in a follow-up

Reply:
Changed to `m_node.args->ForceSetArg("-bind", "");`